### ISSUE TITLE
Give the option of using cupy instead of numpy in processing

### ISSUE THREAD

**Issue by NickGuAI** [nilearn/nilearn#3401]
maybe in imports, add a param check like
```
if os.environ["GPU"] == "enabled":
  import cupy as np
else:
  import numpy as np
```

The current impl is way too slow when running operations at scale.
I.E I was trying to extract raw signal from ~150-500 ROIs, with 48 scans of 1M voxels each.
it has been 7 hours and it's still running....
and I have 4 studies, the other ones have 180-360 scans w/ ~1M voxels each...

fwiw, this is what I've been using...
```

for atlas in atlases:
   for roi in single_roi_masks:
        nifti_masker = NiftiMasker(
            mask_img=f'{single_roi_dir}/{roi}',
            standardize=False, mask_strategy='epi',
            memory=f"{raw_sig_dir}/nilearn_cache")
        for sub, session, cond in experiments:
           imgs = getimgs(sub, session, cond)
            raw_signal = masker.fit_transform(
                        imgs, confounds=confounds
                    )
            np.savetxt(fn, raw_signal, delimiter=",")

```

**Comment by jeromedockes** [comment]
Thanks! and how long did it take to run with the change you suggest?

**Comment by bthirion** [comment]
It would be interesting to see in which step the time is spent: It is likely in image resampling ?

**Comment by DasDominus** [comment]
I have been trying to crack this down... the filter&extract actually takes all the time. ~80s. resampling takes ~0.4s
GPU is a bit tricky as it turns out. pandas needs to be replaced with cudf, and I've encountered some compatibility issues....

**Comment by ymzayek** [comment]
Related to WIPs: #3398 and #3399. We are working on profiling the time and memory consumption of the maskers and it seems that most resources are spend on either resampling or the timeseries extraction step that uses `masking.apply_mask`. So I presume in this case it is the latter. We have yet to identify the most relevant lower level functions.

**Comment by DasDominus** [comment]
Thanks @ymzayek ! Did not know that. Resampling does contribute to majority of the time. Seems it's best to do this first, save the resampled images to disk, and do the transform. I've done profiling for fit_transform:

```
8439844 function calls (7802739 primitive calls) in 306.494 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      802   74.300    0.093   74.300    0.093 {method 'sort' of 'numpy.ndarray' objects}
      800   51.151    0.064  183.682    0.230 resampling.py:247(_resample_one_img)
   633356   37.545    0.000   37.545    0.000 {method 'decompress' of 'zlib.Decompress' objects}
       19   36.575    1.925   36.575    1.925 {built-in method io.open}
     2400   33.060    0.014   33.060    0.014 {built-in method scipy.ndimage._nd_image.spline_filter1d}
      881   17.997    0.020   17.997    0.020 {built-in method builtins.sorted}
635234/1872   17.702    0.000   73.326    0.039 {method 'read' of '_io.BufferedReader' objects}
   633354    7.433    0.000   69.915    0.000 _compression.py:66(readinto)
   633362    7.146    0.000    7.146    0.000 {built-in method zlib.crc32}
       20    4.150    0.207    4.150    0.207 {built-in method posix.stat}
        3    2.897    0.966   74.049   24.683 volumeutils.py:450(array_from_file)
      802    2.204    0.003   78.678    0.098 arraysetops.py:320(_unique1d)
      814    2.164    0.003    2.164    0.003 {method 'flatten' of 'numpy.ndarray' objects}
   633356    1.891    0.000   62.101    0.000 gzip.py:449(read)
        2    1.656    0.828   71.151   35.576 {method 'readinto' of '_io._BufferedIOBase' objects}
      820    1.424    0.002    1.424    0.002 {built-in method numpy.zeros}
        1    1.035    1.035    1.035    1.035 {method 'peek' of '_io.BufferedReader' objects}
   633356    0.834    0.000    8.048    0.000 gzip.py:500(_add_read_data)
      800    0.811    0.001    0.811    0.001 {built-in method scipy.ndimage._nd_image.zoom_shift}
      889    0.739    0.001    0.739    0.001 {method 'reduce' of 'numpy.ufunc' objects}
       99    0.729    0.007    0.729    0.007 {method 'update' of '_hashlib.HASH' objects}
        1    0.671    0.671    0.671    0.671 {method 'read' of '_io.TextIOWrapper' objects}
3173011/3172929    0.318    0.000    0.318    0.000 {built-in method builtins.len}
        3    0.309    0.103    0.309    0.103 {built-in method gc.collect}
        1    0.305    0.305    0.368    0.368 npyio.py:1202(savetxt)
   633368    0.268    0.000   14.559    0.000 gzip.py:85(read)
        1    0.193    0.193  260.126  260.126 memory.py:771(call)
   633354    0.192    0.000    0.192    0.000 {method 'cast' of 'memoryview' objects}
        1    0.064    0.064  184.402  184.402 resampling.py:309(resample_img)
      802    0.061    0.000   78.744    0.098 arraysetops.py:138(unique)
```

**Comment by DasDominus** [comment]
With separated resample, I can save the time by ~15%. with GPU we could cut another ~30% ish (suppose vram holds).

**Comment by ymzayek** [comment]
Thanks @DasDominus this is great! Keep us updated about whether you can get the GPU option with CuPy to work seamlessly. However, I think optimizing the maskers can be done on separate levels. For example, we plan to better document this behavior and what can be done on the user side to cut down the time, such as separating the resampling step and not upsampling the data (see #3397). And profiling time like you did as well as memory will give us a better sense if anything can be done on the nilearn side to optimize the maskers code.

**Comment by htwangtw** [comment]
Thanks for bringing this up!
Just to add that, resampling time can vary a lot depending on how user set the options.
For example, if the data is in 3 mm space, the atlas is in 1 mm space will take a really long time to resample comparing to one in 2mm space. This is not uncommon for atlas made from surface to volumetric data.

**Comment by man-shu** [comment]
This page from scikit-learn docs might be relevant: https://scikit-learn.org/stable/modules/array_api.html

**Comment by man-shu** [comment]
Reopening this because #5958 only demos how to use sklearn estimators to use GPU, but does not yet implement nilearn native GPU support
